src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -14,55 +14,75 @@
 import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 
-/** Servlet to store and get retrieve information. */
+/** Servlet to store and get retrieve user information. */
 @WebServlet("/user")
 public final class UserServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get the id from the query string.
-    long id = request.getParameter("id");
+    UserService userService = UserServiceFactory.getUserService();
 
+    // Get the key from the query string.
+    String keyString = request.getParameter("key");
+    Key userKey = KeyFactory.stringToKey(keyString);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Entity userEntity = datastore.get(id);
-    String email = userEntity.getProperty("email");
-    String username = userEntity.getProperty("username");
-    User user = new User(id, email, username);
 
-    // Convert to JSON and send it as the response.
-    Gson gson = new Gson();
+    try{
+      Entity userEntity = datastore.get(userKey);
+      String email = (String) userEntity.getProperty("email");
+      String username = (String) userEntity.getProperty("username");
+      User user = new User(userKey.getName(), email, username);
 
-    response.setContentType("application/json");
-    response.getWriter().println(gson.toJson(user));
+      // Convert to JSON and send it as the response.
+      Gson gson = new Gson();
+
+      response.setContentType("application/json");
+      response.getWriter().println(gson.toJson(user));
+    }
+    catch(EntityNotFoundException e) {
+      throw new IOException("Entity not found.");
+    }
   }
 
+  // Create or update a user
   @Override
-  public void doPost(HttpServletRequeest request, HttpServletResponse response) throws IOException {   
-    long id = request.getParameter("id");
-    String email = request.getParameter("email");
-    String username = request.getParameter("username");
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    String email = userService.getCurrentUser().getEmail();
+    String id = userService.getCurrentUser().getUserId();
+    String username = request.getParameter("username-input");
+
+    Key userKey = KeyFactory.createKey("User", id);
 
     // Create a new User entity with data from the request.
-    Entity userEntity = new Entity("User");
-    userEntity.setProperty("id", id);
+    Entity userEntity = new Entity(userKey);
```

Comment:
I think you should get the corresponding user from data store and update on only non-empty values from the request. Creating a new entity would defeat the purpose.

Reply:
The way I planned to do it (and eventually did in later PRs), there are never any empty values in the request. The profile page's edit form will be populated with the old values, and the user can edit what they wish and then submit the whole
form. The documentation also states that "The Datastore API does not distinguish between creating a new entity and updating an existing one. If the object's key represents an entity that already exists, the put() method overwrites the existing entity." https://cloud.google.com/appengine/docs/standard/java/datastore/creating-entities#Updating_entities